include/tapkee/defines/methods.hpp

```diff
@@ -119,6 +34,107 @@ template <typename M> struct Method
     const char* name_;
 };
 
+typedef std::tuple<bool, bool, bool> KFDTraits;
+
+struct DimensionReductionMethod : public Method<DimensionReductionMethod>
+{
+    DimensionReductionMethod(const char* n, const KFDTraits& traits)
+        : Method<DimensionReductionMethod>(n)
+        , needs_kernel(std::get<0>(traits))
+        , needs_distance(std::get<1>(traits))
+        , needs_features(std::get<2>(traits))
+    {
+    }
+    using Method<DimensionReductionMethod>::operator==;
+    bool needs_kernel;
+    bool needs_distance;
+    bool needs_features;
+};
+
+static const KFDTraits RequiresKernel{true, false, false};
```

Comment:
Maybe here consider `std::bitset<3>`, it will look a bit more concise, I think particularly below in the lookup with `traits.test(i)` instead of `std::get<i>(traits)`. Although for readability or expressiveness below, probably a struct with three names fields would be preferable. I think all three are good and probably prefer the bitset one most :-)